File: src/SIL.Machine.TS/src/translation/error-correction-model.ts
import { EcmScoreInfo } from './ecm-score-info';
import { EditOperation } from './edit-operation';
import { SegmentEditDistance } from './segment-edit-distance';
import { TranslationResultBuilder } from './translation-result-builder';

export class ErrorCorrectionModel {
  private readonly segmentEditDistance: SegmentEditDistance = new SegmentEditDistance();

  constructor() {
    this.setErrorModelParameters(128, 0.8, 1, 1, 1);
  }

  setErrorModelParameters(
    vocSize: number,
    hitProb: number,
    insFactor: number,
    substFactor: number,
    delFactor: number
  ): void {
    let e: number;
    if (vocSize === 0) {
      e = (1 - hitProb) / (insFactor + substFactor + delFactor);
    } else {
      e = (1 - hitProb) / (insFactor * vocSize + substFactor * (vocSize - 1) + delFactor);
    }

    const insProb = e * insFactor;
    const substProb = e * substFactor;
    const delProb = e * delFactor;

    this.segmentEditDistance.hitCost = -Math.log(hitProb);
    this.segmentEditDistance.insertionCost = -Math.log(insProb);
    this.segmentEditDistance.substitutionCost = -Math.log(substProb);
    this.segmentEditDistance.deletionCost = -Math.log(delProb);
  }

  setupInitialEsi(initialEsi: EcmScoreInfo): void {
    const score = this.segmentEditDistance.compute([], []);
    initialEsi.scores.length = 0;
    initialEsi.scores.push(score);
    initialEsi.operations.length = 0;
  }

  setupEsi(esi: EcmScoreInfo, prevEsi: EcmScoreInfo, word: string): void {
    const score = this.segmentEditDistance.compute([word], []);
    esi.scores.length = 0;
    esi.scores.push(prevEsi.scores[0] + score);
    esi.operations.length = 0;
    esi.operations.push(EditOperation.None);
  }

  extendInitialEsi(initialEsi: EcmScoreInfo, prevInitialEsi: EcmScoreInfo, prefixDif: string[]): void {
    this.segmentEditDistance.incrComputePrefixFirstRow(initialEsi.scores, prevInitialEsi.scores, prefixDif);
  }

  extendEsi(
    esi: EcmScoreInfo,
    prevEsi: EcmScoreInfo,
    word: string,
    prefixDif: string[],
    isLastWordComplete: boolean
  ): void {
    const ops = this.segmentEditDistance.incrComputePrefix(
      esi.scores,
      prevEsi.scores,
      word,
      prefixDif,
      isLastWordComplete
    );
    esi.operations.push(...ops);
  }

  correctPrefix(
    builder: TranslationResultBuilder,
    uncorrectedPrefixLen: number,
    prefix: string[],
    isLastWordComplete: boolean
  ): number {
    if (uncorrectedPrefixLen === 0) {
      for (const w of prefix) {
        builder.appendWord(w);
      }
      return prefix.length;
    }

    const result = this.segmentEditDistance.computePrefix(
      builder.words.slice(0, uncorrectedPrefixLen),
      prefix,
      isLastWordComplete,
      false
    );
    return builder.correctPrefix(result.wordOps, result.charOps, prefix, isLastWordComplete);
  }
}

File: src/SIL.Machine/Translation/ErrorCorrectionModel.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SIL.Machine.Translation
{
    public class ErrorCorrectionModel
    {
        private readonly SegmentEditDistance _segmentEditDistance;

        public ErrorCorrectionModel()
        {
            _segmentEditDistance = new SegmentEditDistance();
            SetErrorModelParameters(128, 0.8, 1, 1, 1);
        }

        public void SetErrorModelParameters(
            int vocSize,
            double hitProb,
            double insFactor,
            double substFactor,
            double delFactor
        )
        {
            double e;
            if (vocSize == 0)
                e = (1 - hitProb) / (insFactor + substFactor + delFactor);
            else
                e = (1 - hitProb) / ((insFactor * vocSize) + (substFactor * (vocSize - 1)) + delFactor);

            double insProb = e * insFactor;
            double substProb = e * substFactor;
            double delProb = e * delFactor;

            _segmentEditDistance.HitCost = -Math.Log(hitProb);
            _segmentEditDistance.InsertionCost = -Math.Log(insProb);
            _segmentEditDistance.SubstitutionCost = -Math.Log(substProb);
            _segmentEditDistance.DeletionCost = -Math.Log(delProb);
        }

        public void SetupInitialEsi(EcmScoreInfo initialEsi)
        {
            double score = _segmentEditDistance.Compute(new string[0], new string[0]);
            initialEsi.Scores.Clear();
            initialEsi.Scores.Add(score);
            initialEsi.Operations.Clear();
        }

        public void SetupEsi(EcmScoreInfo esi, EcmScoreInfo prevEsi, string word)
        {
            double score = _segmentEditDistance.Compute(new string[] { word }, new string[0]);
            esi.Scores.Clear();
            esi.Scores.Add(prevEsi.Scores[0] + score);
            esi.Operations.Clear();
            esi.Operations.Add(EditOperation.None);
        }

        public void ExtendInitialEsi(EcmScoreInfo initialEsi, EcmScoreInfo prevInitialEsi, string[] prefixDiff)
        {
            _segmentEditDistance.IncrComputePrefixFirstRow(initialEsi.Scores, prevInitialEsi.Scores, prefixDiff);
        }

        public void ExtendEsi(
            EcmScoreInfo esi,
            EcmScoreInfo prevEsi,
            string word,
            string[] prefixDiff,
            bool isLastWordComplete
        )
        {
            IEnumerable<EditOperation> ops = _segmentEditDistance.IncrComputePrefix(
                esi.Scores,
                prevEsi.Scores,
                word,
                prefixDiff,
                isLastWordComplete
            );
            foreach (EditOperation op in ops)
                esi.Operations.Add(op);
        }

        public int CorrectPrefix(
            TranslationResultBuilder builder,
            int uncorrectedPrefixLen,
            string[] prefix,
            bool isLastWordComplete
        )
        {
            if (uncorrectedPrefixLen == 0)
            {
                foreach (string w in prefix)
                    builder.AppendToken(w, TranslationSources.Prefix, -1);
                return prefix.Length;
            }

            IEnumerable<EditOperation> wordOps,
                charOps;
            _segmentEditDistance.ComputePrefix(
                builder.TargetTokens.Take(uncorrectedPrefixLen).ToArray(),
                prefix,
                isLastWordComplete,
                false,
                out wordOps,
                out charOps
            );
            return builder.CorrectPrefix(wordOps, charOps, prefix, isLastWordComplete);
        }
    }
}
